File: sdk/python/bunqueue/queue.py
"""Queue: producer + control API over the bunqueue TCP protocol.

Feature parity with the TypeScript client's Queue (TCP mode). Query and
admin surfaces live in :mod:`queue_query` / :mod:`queue_admin` mixins.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Sequence

from .connection import Connection, TlsOption, _compact
from .job import Job
from .options import job_options, job_payload
from .queue_admin import QueueAdminOps
from .queue_query import QueueQueryOps


class Queue(QueueQueryOps, QueueAdminOps):
    """Producer and management client for one named queue."""

    def __init__(
        self,
        name: str,
        host: str = "localhost",
        port: int = 6789,
        token: Optional[str] = None,
        tls: TlsOption = None,
        connection: Optional[Connection] = None,
        command_timeout: float = 10.0,
        prefix_key: str = "",
    ) -> None:
        self.name = prefix_key + name
        self.connection = connection or Connection(
            host=host, port=port, token=token, tls=tls, command_timeout=command_timeout
        )
        self._owns_connection = connection is None

    # --------------------------------------------------------------- produce

    def add(self, name: str, data: Any = None, **options: Any) -> Job:
        """Add a job; returns a Job stub carrying the assigned id."""
        payload = job_payload(name, data)
        opts = job_options(**options)
        response = self.connection.call({"cmd": "PUSH", "queue": self.name, "data": payload, **opts})
        return Job({"id": response.get("id"), "queue": self.name, "data": payload, **opts}, self.connection)

    def add_bulk(self, jobs: Sequence[Dict[str, Any]]) -> List[str]:
        """Add many jobs in one round-trip.

        Each entry: ``{"name": str, "data": Any, ...options}`` with the same
        option names accepted by :meth:`add`.
        """
        inputs = []
        for entry in jobs:
            entry = dict(entry)
            name = entry.pop("name")
            data = entry.pop("data", None)
            opts = job_options(**entry)
            # PUSHB entries are typed JobInput, whose custom-id field is
            # `customId` — unlike single PUSH which renames `jobId`->`customId`
            # server-side. Without this rename the batch custom id is dropped
            # (idempotency/getJobByCustomId broken).
            if "jobId" in opts:
                opts["customId"] = opts.pop("jobId")
            inputs.append({"data": job_payload(name, data), **opts})
        response = self.connection.call({"cmd": "PUSHB", "queue": self.name, "jobs": inputs})
        return [str(job_id) for job_id in response.get("ids", [])]

    # --------------------------------------------------------------- control

    def pause(self) -> None:
        self.connection.call({"cmd": "Pause", "queue": self.name})

    def resume(self) -> None:
        self.connection.call({"cmd": "Resume", "queue": self.name})

    def is_paused(self) -> bool:
        response = self.connection.call({"cmd": "IsPaused", "queue": self.name})
        return bool(response.get("paused"))

    def drain(self) -> None:
        self.connection.call({"cmd": "Drain", "queue": self.name})

    def obliterate(self) -> None:
        self.connection.call({"cmd": "Obliterate", "queue": self.name})

    def clean(self, grace_ms: int, limit: Optional[int] = None, state: Optional[str] = None) -> int:
        response = self.connection.call(
            _compact({"cmd": "Clean", "queue": self.name, "grace": grace_ms, "state": state, "limit": limit})
        )
        return int(response.get("count", 0))

    def remove(self, job_id: str) -> None:
        self.connection.call({"cmd": "Cancel", "id": job_id})

    # `cancel` kept as an alias of remove (bunqueue-native name)
    cancel = remove

    def discard(self, job_id: str) -> None:
        self.connection.call({"cmd": "Discard", "id": job_id})

    def promote_job(self, job_id: str) -> None:
        self.connection.call({"cmd": "Promote", "id": job_id})

    def promote_jobs(self, count: Optional[int] = None) -> int:
        response = self.connection.call(
            _compact({"cmd": "PromoteJobs", "queue": self.name, "count": count})
        )
        return int(response.get("count", 0))

    def retry_job(self, job_id: str) -> None:
        """BullMQ contract: failed -> waiting (MoveToWait over TCP)."""
        self.connection.call({"cmd": "MoveToWait", "id": job_id})

    def retry_jobs(self, state: str = "failed", count: Optional[int] = None) -> int:
        # `count` caps how many DLQ entries are retried (server >= 2.8.29);
        # older servers ignore it and retry the whole DLQ (forward-compatible).
        if state == "failed":
            response = self.connection.call(
                _compact({"cmd": "RetryDlq", "queue": self.name, "count": count})
            )
        elif state == "completed":
            response = self.connection.call({"cmd": "RetryCompleted", "queue": self.name})
        else:
            raise ValueError("state must be 'failed' or 'completed'")
        return int(response.get("count", 0))

    def retry_completed(self, job_id: Optional[str] = None) -> int:
        response = self.connection.call(
            _compact({"cmd": "RetryCompleted", "queue": self.name, "id": job_id})
        )
        return int(response.get("count", 0))

    # ----------------------------------------------------- job state changes

    def update_job_progress(self, job_id: str, progress: int, message: Optional[str] = None) -> None:
        self.connection.call(
            _compact({"cmd": "Progress", "id": job_id, "progress": progress, "message": message})
        )

    def update_job_data(self, job_id: str, data: Any) -> None:
        self.connection.call({"cmd": "Update", "id": job_id, "data": data})

    def change_job_priority(self, job_id: str, priority: int, lifo: Optional[bool] = None) -> None:
        self.connection.call(
            _compact({"cmd": "ChangePriority", "id": job_id, "priority": priority, "lifo": lifo})
        )

    def change_job_delay(self, job_id: str, delay_ms: int) -> None:
        self.connection.call({"cmd": "ChangeDelay", "id": job_id, "delay": delay_ms})

    def move_job_to_wait(self, job_id: str) -> None:
        self.connection.call({"cmd": "MoveToWait", "id": job_id})

    def move_job_to_delayed(self, job_id: str, delay_ms: int) -> None:
        self.connection.call({"cmd": "MoveToDelayed", "id": job_id, "delay": delay_ms})

    def extend_job_lock(self, job_id: str, token: str, duration_ms: int) -> None:
        self.connection.call(
            {"cmd": "ExtendLock", "id": job_id, "token": token, "duration": duration_ms}
        )

    def move_job_to_completed(self, job_id: str, result: Any = None, token: Optional[str] = None) -> None:
        self.connection.call(_compact({"cmd": "ACK", "id": job_id, "result": result, "token": token}))

    def move_job_to_failed(self, job_id: str, error: str, token: Optional[str] = None) -> None:
        self.connection.call(_compact({"cmd": "FAIL", "id": job_id, "error": error, "token": token}))

    # ------------------------------------------------------------- lifecycle

    def ping(self) -> bool:
        return self.connection.ping()

    def wait_until_ready(self) -> None:
        self.connection.connect()

    def close(self) -> None:
        if self._owns_connection:
            self.connection.close()

    def disconnect(self) -> None:
        """BullMQ-parity alias for :meth:`close`."""
        self.close()

    def __enter__(self) -> "Queue":
        return self

    def __exit__(self, *exc: Any) -> None:
        self.close()

File: sdk/elixir/lib/bunqueue/queue.ex
defmodule Bunqueue.Queue do
  @moduledoc "Producing API for a named bunqueue queue."

  alias Bunqueue.{Connection, Job, Options}

  @enforce_keys [:name, :connection]
  defstruct [:name, :connection, owns_connection: false]

  @type t :: %__MODULE__{
          name: String.t(),
          connection: GenServer.server(),
          owns_connection: boolean()
        }

  @spec new(String.t(), GenServer.server(), boolean()) :: t()
  def new(name, connection, owns_connection \\ false) do
    %__MODULE__{name: name, connection: connection, owns_connection: owns_connection}
  end

  @spec add(t(), String.t(), term(), keyword() | map()) ::
          {:ok, Job.t()} | {:error, Exception.t()}
  def add(queue, name, data \\ %{}, options \\ []) do
    payload = job_payload(name, data)

    command =
      %{"cmd" => "PUSH", "queue" => queue.name, "data" => payload}
      |> Map.merge(Options.job(options))

    with {:ok, response} <- call(queue, command) do
      raw = %{"id" => response["id"], "queue" => queue.name, "data" => payload}
      {:ok, Job.from_wire(raw, queue.connection)}
    end
  end

  @spec add!(t(), String.t(), term(), keyword() | map()) :: Job.t()
  def add!(queue, name, data \\ %{}, options \\ []) do
    case add(queue, name, data, options) do
      {:ok, job} -> job
      {:error, error} -> raise error
    end
  end

  @spec add_bulk(t(), [map()]) :: {:ok, [String.t()]} | {:error, Exception.t()}
  def add_bulk(queue, entries) do
    jobs =
      Enum.map(entries, fn entry ->
        name = fetch(entry, :name)
        data = get(entry, :data, %{})
        options = get(entry, :opts, get(entry, :options, []))

        %{"data" => job_payload(name, data)}
        |> Map.merge(Options.job(options, true))
      end)

    with {:ok, response} <-
           call(queue, %{"cmd" => "PUSHB", "queue" => queue.name, "jobs" => jobs}) do
      {:ok, response["ids"] || []}
    end
  end

  @spec hello(t()) :: {:ok, map()} | {:error, Exception.t()}
  def hello(queue) do
    call(queue, %{
      "cmd" => "Hello",
      "protocolVersion" => 2,
      "capabilities" => []
    })
  end

  @spec ping(t()) :: {:ok, boolean()} | {:error, Exception.t()}
  def ping(queue) do
    with {:ok, %{"data" => data}} <- call(queue, %{"cmd" => "Ping"}) do
      {:ok, data["pong"] == true}
    end
  end

  defdelegate get_job(queue, id), to: Bunqueue.QueueQuery
  defdelegate get_job_by_custom_id(queue, custom_id), to: Bunqueue.QueueQuery
  defdelegate get_jobs(queue), to: Bunqueue.QueueQuery
  defdelegate get_jobs(queue, state), to: Bunqueue.QueueQuery
  defdelegate get_jobs(queue, state, offset), to: Bunqueue.QueueQuery
  defdelegate get_jobs(queue, state, offset, limit), to: Bunqueue.QueueQuery
  defdelegate get_state(queue, id), to: Bunqueue.QueueQuery
  defdelegate get_result(queue, id), to: Bunqueue.QueueQuery
  defdelegate get_progress(queue, id), to: Bunqueue.QueueQuery
  defdelegate get_job_counts(queue), to: Bunqueue.QueueQuery
  defdelegate count(queue), to: Bunqueue.QueueQuery
  defdelegate wait_for_job(queue, id, timeout_ms), to: Bunqueue.QueueQuery
  defdelegate get_logs(queue, id), to: Bunqueue.QueueQuery
  defdelegate get_logs(queue, id, start), to: Bunqueue.QueueQuery
  defdelegate get_logs(queue, id, start, finish), to: Bunqueue.QueueQuery
  defdelegate pause(queue), to: Bunqueue.QueueControl
  defdelegate resume(queue), to: Bunqueue.QueueControl
  defdelegate is_paused(queue), to: Bunqueue.QueueControl
  defdelegate drain(queue), to: Bunqueue.QueueControl
  defdelegate clean(queue, grace_ms, limit, state), to: Bunqueue.QueueControl
  defdelegate obliterate(queue), to: Bunqueue.QueueControl
  defdelegate cancel(queue, id), to: Bunqueue.QueueControl
  defdelegate discard(queue, id), to: Bunqueue.QueueControl
  defdelegate promote(queue, id), to: Bunqueue.QueueControl
  defdelegate promote_jobs(queue), to: Bunqueue.QueueControl
  defdelegate retry_job(queue, id), to: Bunqueue.QueueControl
  defdelegate move_to_delayed(queue, id, delay_ms), to: Bunqueue.QueueControl
  defdelegate change_priority(queue, id, priority), to: Bunqueue.QueueControl
  defdelegate change_delay(queue, id, delay_ms), to: Bunqueue.QueueControl
  defdelegate update(queue, id, data), to: Bunqueue.QueueControl
  defdelegate update_parent(queue, child_id, parent_id), to: Bunqueue.QueueControl
  defdelegate dlq(queue), to: Bunqueue.QueueAdmin
  defdelegate dlq(queue, count), to: Bunqueue.QueueAdmin
  defdelegate retry_dlq(queue), to: Bunqueue.QueueAdmin
  defdelegate retry_dlq(queue, job_id), to: Bunqueue.QueueAdmin
  defdelegate retry_dlq(queue, job_id, count), to: Bunqueue.QueueAdmin
  defdelegate purge_dlq(queue), to: Bunqueue.QueueAdmin
  defdelegate upsert_scheduler(queue, id, repeat), to: Bunqueue.QueueAdmin
  defdelegate upsert_scheduler(queue, id, repeat, template), to: Bunqueue.QueueAdmin
  defdelegate get_scheduler(queue, id), to: Bunqueue.QueueAdmin
  defdelegate list_schedulers(queue), to: Bunqueue.QueueAdmin
  defdelegate remove_scheduler(queue, id), to: Bunqueue.QueueAdmin
  defdelegate set_rate_limit(queue, limit), to: Bunqueue.QueueAdmin
  defdelegate set_rate_limit(queue, limit, options), to: Bunqueue.QueueAdmin
  defdelegate clear_rate_limit(queue), to: Bunqueue.QueueAdmin
  defdelegate set_concurrency(queue, limit), to: Bunqueue.QueueAdmin
  defdelegate clear_concurrency(queue), to: Bunqueue.QueueAdmin

  @spec close(t()) :: :ok
  def close(%{owns_connection: true, connection: connection}), do: Connection.close(connection)
  def close(_queue), do: :ok

  @doc false
  def call(queue, command, timeout \\ nil),
    do: Connection.call(queue.connection, command, timeout)

  @doc false
  def job_payload(name, data) when is_map(data) do
    normalized = Map.new(data, fn {key, value} -> {to_string(key), value} end)
    Map.merge(%{"name" => name}, normalized)
  end

  def job_payload(name, nil), do: %{"name" => name}
  def job_payload(name, data), do: %{"name" => name, "payload" => data}

  defp fetch(map, key) do
    case Map.fetch(map, key) do
      {:ok, value} -> value
      :error -> Map.fetch!(map, to_string(key))
    end
  end

  defp get(map, key, default) do
    Map.get(map, key, Map.get(map, to_string(key), default))
  end
end
